src/dag/chunk.ts

```diff
@@ -2,32 +2,36 @@ import {assertString} from '../asserts';
 import {Hash} from '../hash';
 import type {Value} from '../kv/store';
 
-type Refs = string[];
+type Refs = readonly string[];
 
 // TODO(arv): Make this class take a type parameter for the data type?
-export class Chunk {
+export class Chunk<V extends Value = Value> {
   readonly hash: string;
-  readonly data: Value;
+  readonly data: V;
   /**
    * Meta is an array of refs. If there are no refs we do not write a meta
    * chunk.
    */
   readonly meta: Refs;
 
-  private constructor(hash: string, data: Value, meta: Refs = []) {
+  private constructor(hash: string, data: V, meta: Refs = []) {
     this.hash = hash;
     this.data = data;
     this.meta = meta;
   }
 
-  static new(data: Value, refs: string[]): Chunk {
+  static new<V extends Value = Value>(data: V, refs: Refs): Chunk<V> {
     // Use hash of JSON stringified data if a JSONValue is passed.
-    const sum = ArrayBuffer.isView(data) ? data : JSON.stringify(data);
+    const sum = data instanceof Uint8Array ? data : JSON.stringify(data);
```

Comment:
I see --you can use type to discriminate all the way up the stack. OK.

I still feel nervous having this code mixed everywhere into the codebase. Feels more robust to have an entirely separate codepath that just knows how to read the old data into the new data structures which can be tested in isolation. WDYT?